test/ip_constexpr_test.cpp

```diff
@@ -0,0 +1,76 @@
+//
+// ip_constexpr_test.cpp
+//
+// Copyright 2025 Mathias Stearn
+//
+// Distributed under the Boost Software License, Version 1.0.
+// See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt
+//
+
+#include <boost/intrusive_ptr.hpp>
+
+#ifndef BOOST_SP_NO_CXX20_CONSTEXPR
+
+struct dummy {
+    // no-ops, so safe on pointers to static constexpr variables
+    friend constexpr void intrusive_ptr_add_ref(const dummy *) {}
+    friend constexpr void intrusive_ptr_release(const dummy *) {}
+};
+static constexpr dummy d;
+
+struct subdummy : dummy {};
+
+// Test that basic operations work at compile time.
+static_assert(bool(boost::intrusive_ptr(&d)));
+static_assert(!bool(boost::intrusive_ptr<dummy>(nullptr)));
+static_assert(!bool(boost::intrusive_ptr<dummy>()));
+static_assert(!bool(boost::intrusive_ptr<dummy>(boost::intrusive_ptr<subdummy>())));
+static_assert(&*boost::intrusive_ptr(&d));
+static_assert(boost::intrusive_ptr(&d).operator->());
+static_assert(boost::intrusive_ptr<dummy>() == nullptr);
+static_assert(boost::intrusive_ptr<dummy>() == boost::intrusive_ptr<dummy>(nullptr));
+static_assert(boost::intrusive_ptr<dummy>() != boost::intrusive_ptr(&d));
+static_assert(boost::intrusive_ptr(&d) != nullptr);
+static_assert(boost::intrusive_ptr(&d) == boost::intrusive_ptr(&d));
+static_assert(boost::intrusive_ptr(&d) == boost::intrusive_ptr(&d).get());
+static_assert(boost::intrusive_ptr(&d) == boost::intrusive_ptr(&d).detach());
+static_assert(!(boost::intrusive_ptr(&d) < boost::intrusive_ptr(&d)));
+static_assert(boost::get_pointer(boost::intrusive_ptr(&d)));
+static_assert(boost::static_pointer_cast<const dummy>(boost::intrusive_ptr(&d)));
+static_assert(boost::const_pointer_cast<const dummy>(boost::intrusive_ptr(&d)));
+static_assert(boost::dynamic_pointer_cast<const dummy>(boost::intrusive_ptr(&d)));
+
+constexpr auto lvalue = boost::intrusive_ptr(&d);
+constexpr auto lvalue_convertible = boost::intrusive_ptr<const subdummy>();
+static_assert(boost::intrusive_ptr(lvalue));
+static_assert(!boost::intrusive_ptr(lvalue_convertible));
+static_assert(boost::static_pointer_cast<const dummy>(lvalue));
+static_assert(boost::const_pointer_cast<const dummy>(lvalue));
+static_assert(boost::dynamic_pointer_cast<const dummy>(lvalue));
+
+constexpr bool test_swap() {
+    auto p1 = boost::intrusive_ptr(&d);
+    auto p2 = boost::intrusive_ptr(&d);
+    swap(p1, p2);
+    p1.swap(p2);
+    return  true;
+}
+static_assert(test_swap());
+
+constexpr bool test_reset_assign() {
+    auto p1 = boost::intrusive_ptr(&d);
+    p1.reset();
+    p1.reset(&d);
+    p1.reset(&d, true);
+    p1 = boost::intrusive_ptr(&d);
+    p1 = lvalue;
+    p1 = nullptr;
+    p1 = &d;
+    p1 = boost::intrusive_ptr<subdummy>();
+    p1 = lvalue_convertible;
+    return true;
+}
+static_assert(test_reset_assign());
+
+#endif
```

Comment:
The lack of a newline at the end is probably not intentional.

Reply:
Sorry, just realized that my config to add it is workspace specific. Will fix.

PS - it'd be great to add a .clang-format to the repo root so that formatting things are handled automatically. I've had to run "save without formatting" to avoid reflowing the whole file every time I save.